src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5874,201 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
+  ENCODE_NUMBER(buff[i], message_crc, i);
+
+  return i;
+}
+
+#define PAYLOAD_LINE "\n<Payload>\n<Records>\n<Payload>\n"
+#define END_PAYLOAD_LINE "\n</Payload></Records></Payload>"
+
+int RGWS3Select::run_s3select(char*query,char*input,size_t input_length,bool skip_first_line,bool skip_last_line,bool to_aggregate)
+{
+  char * buff = 0;
+  std::string res;
+  int status = 0;
+
+  csv_object s3_csv_object(s3select_syntax ,query ,input ,input_length ,skip_first_line ,skip_last_line ,to_aggregate );
+  if (s3select_syntax->get_error_description().empty() == false)
+  {
+    res.append(PAYLOAD_LINE);
+    res.append(s3select_syntax->get_error_description());
+    status = -1;
+  }
+  else
+  {
+    res.append(PAYLOAD_LINE);
+    status = s3_csv_object.run_s3select_on_object(res);
+    if(status<0)
+    {
+      res.append(s3_csv_object.get_error_description());
+    }
+  }
+
+  if (res.size() > strlen(PAYLOAD_LINE))
+  {
+    res.append(END_PAYLOAD_LINE);
+
+    buff = (char *)malloc(res.size() + 1000);
+    int buff_len = create_message(res.c_str(), buff);
+
+    s->formatter->write_bin_data(buff, buff_len);
+    if (op_ret < 0)
+      return op_ret;
+  }
+  rgw_flush_formatter_and_reset(s, s->formatter);
+
+  if(buff) free(buff);
+
+  return status;
+}
+
+int RGWS3Select::send_response_data(bufferlist &bl, off_t ofs, off_t len)
+{
+  if (len == 0)
+    return 0;
+
+  if (chunk_number == 0)
+  {
+    if (op_ret < 0)
+      set_req_state_err(s, op_ret);
+    dump_errno(s);
+  }
+  // Explicitly use chunked transfer encoding so that we can stream the result
+  // to the user without having to wait for the full length of it.
+  if (chunk_number == 0)
+    end_header(s, this, "application/xml", CHUNKED_TRANSFER_ENCODING);
+
+#define LINE_DELIMITER '\n' //TODO should be dynamic
+#define GT "&gt;"
+#define LT "&lt;"
+  if (m_s3select_query.find(GT) != std::string::npos)
+    boost::replace_all(m_s3select_query, GT, ">");
+  if (m_s3select_query.find(LT) != std::string::npos)
+    boost::replace_all(m_s3select_query, LT, "<");
+
+  size_t _qs = m_s3select_query.find("<Expression>", 0) + strlen("<Expression>");
+  size_t _qe = m_s3select_query.find("</Expression>", _qs);
+
+  std::string query = m_s3select_query.substr(_qs, _qe - _qs);
+
+  std::string res;
+  std::string tmp_buff;
+  std::string merge_line;
+  u_int32_t skip_last_bytes = 0;
+  bool skip_first_line = false;
+  m_processed_bytes += len;
+  int status = 0;
+
+  if (m_previous_line)
+  { //if previous broken line exist , merge it to current chunk
+    char *p_obj_chunk = (char *)bl.c_str();
+    while (*p_obj_chunk != LINE_DELIMITER)
+      p_obj_chunk++;
+
+    tmp_buff.assign((char *)bl.c_str(), (char *)bl.c_str() + (p_obj_chunk - bl.c_str()));
+    merge_line = m_last_line + tmp_buff + LINE_DELIMITER;
+    m_previous_line = false;
+    skip_first_line = true;
+
+    status = run_s3select((char *)query.c_str(), (char *)merge_line.c_str(), strlen(merge_line.c_str()), false, false, false);
```

Comment:
please avoid C-style casts, if you want to do const_cast, be explicit

Reply:
the signature for /run_s3select/ is not correct , will be modified.